backend/server/src/derivation.rs

```diff
@@ -0,0 +1,142 @@
+    /// derivation only needs to fully build once, later commits may still include this
+    /// derivation but do not trigger a new build.
+    pub commit: gix_hash::ObjectId,
+
+    /// The Nix command that was used to build this derivation.
+    pub build_command: DrvBuildCommand,
+}
+
+/// Command used to build the derivation.
+pub enum DrvBuildCommand {
+    /// Nix 1.0 style command. Fields are TBD.
+    Nix1,
+}
+
+/// Build events are emitted whenever a derivation build's state changes.
+pub struct DrvBuildEvent {
+    /// The derivation build this event is associated with.
+    pub build: DrvBuildId,
+
+    /// Monotonically increasing numeric event identifier. Use this to determine the order of
+    /// events. Events with a higher id happened later.
+    ///
+    /// Since this id is global it can be used to retrieve the latest build state for a derivation
+    /// without having to check [DrvBuild::build_attempt].
+    pub event_id: u32,
+
+    /// The build state this event propagates.
+    pub state: DrvBuildState,
+
+    /// The timestamp when this event happened.
+    ///
+    /// This timestamp only has second accuracy, which makes it unsuitable for sorting of build
+    /// events. If for example the build queue is empty, it is not unlikely that a build is
+    /// scheduled ([DrvBuildState::Pending]) and started ([DrvBuildState::Building]) within the
+    /// same second.
+    ///
+    /// Using a higher accuracy timestamp would make it incompatible with SQlite's datetime
+    /// functions which is also not desirable.
+    ///
+    /// Instead, use [event_id] to determine the order of events.
+    pub timestamp: chrono::DateTime<chrono::Utc>,
+}
+
+/// Describes the possible states a derivation build can be in.
+pub enum DrvBuildState {
+    /// Build is scheduled in a queue and has not yet started.
+    Pending,
+    /// Build is currently running.
+    Building,
+    /// Build has completed, either successfully or not.
+    Completed {
+        result: DrvBuildResult,
+
+        // TBD
+        build_log: (),
+    },
+    /// Build was interrupted before it could complete.
+    Interrupted { reason: DrvBuildInterruptionKind },
+}
+
+/// Possible causes for why the derivation build was interrupted.
+pub enum DrvBuildInterruptionKind {
+    /// Build process ran out of memory and was killed by the system.
+    OutOfMemory,
+    /// Build process timed out and was killed by the build scheduler.
+    Timeout,
+    /// Scheduler process performed a graceful shutdown and cancelled the derivation build in the
+    /// process.
+    Cancelled,
+    /// Build process died for unknown reasons, most likely a fault in the build command.
+    ProcessDeath,
+    /// Scheduler process died. The scheduler can infer that this happend by checking for
+    /// derivation builds which do not have the status [DrvBuildStatus::Completed] whilst
+    /// starting.
+    SchedulerDeath,
+}
```

Comment:
hmm, I need to think about how terminal states should be represented as well.

OOM could be thought of as a specific kind of ProcessDeath

Reply:
What do you mean by terminal states?

Yes, OOM is just process death. I encoded it separately because I thought it may be one of the more interesting reasons. I think we will probably adjust this enum as we go and figure out what kind of error states we can actually reliably catch and determine later.

Reply:
fair enough, until there's a beta release, I'll likely have no stability guarantees. We should be free-ish to break schemas until we settle on something reasonable.

Reply:
Sounds good